diff --git a/options/options.c b/options/options.c
index 86e5a7b..7dc5d09 100644
--- a/options/options.c
+++ b/options/options.c
@@ -1040,7 +1040,8 @@ static const struct MPOpts mp_default_opts = {
         [STREAM_SUB] = (char *[]){ "auto", NULL },
     },
     .stream_auto_sel = true,
-    .subs_with_matching_audio = false,
+    .subs_with_matching_audio = true,
+    .subs_fallback = 1,
     .subs_fallback_forced = true,
     .audio_display = 1,
     .audio_output_format = 0,  // AF_FORMAT_UNKNOWN
